src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,458 @@
+  private readonly ldp = 'http://www.w3.org/ns/ldp#';
+  private readonly terms = 'http://purl.org/dc/terms/';
+  private readonly xml = 'http://www.w3.org/2001/XMLSchema#';
+  private readonly stat = 'http://www.w3.org/ns/posix/stat#';
+
+  private readonly predicates = {
+    aType: DataFactory.namedNode(`${this.rdf}type`),
+    modified: DataFactory.namedNode(`${this.terms}modified`),
+    contains: DataFactory.namedNode(`${this.ldp}contains`),
+    mtime: DataFactory.namedNode(`${this.stat}mtime`),
+    size: DataFactory.namedNode(`${this.stat}size`),
+  };
+
+  private readonly objects = {
+    container: DataFactory.namedNode(`${this.ldp}Container`),
+    basicContainer: DataFactory.namedNode(`${this.ldp}BasicContainer`),
+    ldpResource: DataFactory.namedNode(`${this.ldp}Resource`),
+    dateTime: DataFactory.namedNode(`${this.xml}dateTime`),
+  };
+
+  /**
+   * @param base - Will be stripped of all incoming URIs and added to all outgoing ones to find the relative path.
+   * @param root - Root filepath in which the resources and containers will be saved as files and directories.
+   */
+  public constructor(base: string, root: string) {
+    this.base = base;
+    while (this.base.endsWith('/')) {
+      this.base = this.base.slice(0, -1);
+    }
+
+    this.root = root;
+    while (this.root.endsWith('/')) {
+      this.root = this.root.slice(0, -1);
+    }
+
+    this.interactionHandler = new InteractionHandler();
+  }
+
+  /**
+   * Store the incoming data as a file under a file path corresponding to `container.path`,
+   * where slashes correspond to subdirectories.
+   * @param container - The identifier to store the new data under.
+   * @param representation - Data to store. Only File streams are supported.
+   *
+   * @returns The newly generated identifier.
+   */
+  public async addResource(container: ResourceIdentifier, representation: Representation): Promise<ResourceIdentifier> {
+    if (representation.dataType !== DATA_TYPE_BINARY) {
+      throw new UnsupportedMediaTypeHttpError('FileResourceStore only supports binary representations.');
+    }
+    const path = this.parseIdentifier(container);
+    const { slug, linkType, raw } = representation.metadata;
+    let metadata;
+    if (raw && raw.length > 0) {
+      metadata = streamifyArray(raw);
+    }
+
+    const resultingAction = this.interactionHandler.getResultingAction(path, slug, linkType);
+    if (resultingAction.requestAction === RequestAction.CREATE_RESOURCE) {
+      // Create a file for the resource with as filepath the parent container.
+      return this.createFile(`${this.root}${resultingAction.parentContainer}`,
+        resultingAction.newIdentifier,
+        representation.data,
+        path.endsWith('/'),
+        metadata);
+    }
+
+    // Create a new container as subdirectory of the parent container.
+    return this.createContainer(`${this.root}${resultingAction.parentContainer}`,
```

Comment:
I don't understand the advantage of putting this in a separate function and then having to check at that place whether a container or a file has to be created.
(The `createFile` function also contains an additional data parameter.)

Reply:
True, I might have been to hasty. Was my immediate reaction after seeing 2 blocks with practically identical code, but since this is the only location where this is done it can stay like this.